core/src/main/scala/akka/kafka/internal/CommittableSources.scala

```diff
@@ -102,22 +102,39 @@ private[kafka] final class CommittableSubSource[K, V](
     ) {
   override protected def logic(
       shape: SourceShape[(TopicPartition, Source[CommittableMessage[K, V], NotUsed])]
-  ): GraphStageLogic with Control =
-    new SubSourceLogic[K, V, CommittableMessage[K, V]](shape, settings, subscription, getOffsetsOnAssign, onRevoke)
-    with CommittableMessageBuilder[K, V] with MetricsControl {
-      override def metadataFromRecord(record: ConsumerRecord[K, V]): String = _metadataFromRecord(record)
-      override def groupId: String = settings.properties(ConsumerConfig.GROUP_ID_CONFIG)
-      lazy val committer: KafkaAsyncConsumerCommitterRef = {
-        val ec = materializer.executionContext
-        new KafkaAsyncConsumerCommitterRef(consumerActor, settings.commitTimeout)(ec)
+  ): GraphStageLogic with Control = {
+    def factory(shape: SourceShape[CommittableMessage[K, V]],
+                tp: TopicPartition,
+                consumerActor: ActorRef,
+                subSourceStartedCb: AsyncCallback[(TopicPartition, (Control, ActorRef))],
+                subSourceCancelledCb: AsyncCallback[(TopicPartition, Option[ConsumerRecord[K, V]])],
+                actorNumber: Int): MessageSubSourceLogic[K, V, CommittableMessage[K, V]] =
+      new MessageSubSourceLogic[K, V, CommittableMessage[K, V]](shape,
```

Comment:
This could get a named type.

Reply:
Concrete type created `CommittableSubSourceStageLogic`

https://github.com/akka/alpakka-kafka/pull/930#discussion_r340823268